gui/layers/_vectors_layer/model.py

```diff
@@ -0,0 +1,615 @@
+                             '5x5': (5, 5),
+                             '7x7': (7, 7),
+                             '9x9': (9, 9),
+                             '11x11': (11, 11)}
+        self._kernel = self._kernel_dict['1x1']
+
+        # Store underlying data model
+        self._data_types = ('matrix', 'projection')
+        self._data_type = None
+
+        # Save the line style params
+        self._width = width
+        self._color = color
+        self._colors = get_color_names()
+        self._connector_types = ['segments']
+        self._connector = connector
+
+        # averaging and length attributes
+        self._avg_dims = ['1x1', '3x3', '5x5', '7x7', '9x9', '11x11']
+        self._averaging = averaging
+        self._length = length
+        self._avg_observers = []
+        self._len_observers = []
+
+        # update flags
+        self._need_display_update = False
+        self._need_visual_update = False
+
+        # assign vector data and establish default behavior
+        self._raw_dat = None
+        self._original_data = vectors
+        self._current_shape = vectors.shape
+        self._vectors = self._check_vector_type(vectors)
+        self.averaging_bind_to(self._default_avg)
+        self.length_bind_to(self._default_length)
+
+        self._mode = 'pan/zoom'
+        self._mode_history = self._mode
+
+        self.name = 'vectors'
+        self.events.add(mode=Event)
+        self._qt_properties = QtVectorsLayer(self)
+        self._qt_controls = QtVectorsControls(self)
+
+    # ====================== Property getter and setters =====================
+    @property
+    def _original_data(self) -> np.ndarray:
+        return self._raw_dat
+
+    @_original_data.setter
+    def _original_data(self, dat: np.ndarray):
+        """
+        Must preserve data used at construction. Specifically for default
+            averaging/length adjustments
+        averaging/length adjustments recalculate the underlying data
+        :param dat: updated only at construction
+        :return:
+        """
+        if self._raw_dat is None:
+            self._raw_dat = dat
+
+    @property
+    def vectors(self) -> np.ndarray:
+        return self._vectors
+
+    @vectors.setter
+    def vectors(self, vectors: np.ndarray):
+        """
+        Can accept two data types:
+            1) (N, 4) array with elements (x, y, u, v),
+                where x-y are position (center) and u-v are x-y projections of
+                    the vector
+            2) (N, M, 2) array with elements (u, v)
+                where u-v are x-y projections of the vector
+                vector position is one per-pixel in the NxM array
+        :param vectors: ndarray
+        :return:
+        """
+        self._original_data = vectors
+
+        self._vectors = self._check_vector_type(vectors)
+
+        self.viewer._child_layer_changed = True
+        self._refresh()
+
+    def _check_vector_type(self, vectors):
```

Comment:
I feel like this function does more than just a check - it actually also returns the `cood_list` that gets set to the `_vectors` property. Can you rename accordingly